### PR TITLE
Add a component for filtering repositories based on their results

### PR DESCRIPTION
**Pull request by robertbrignull**
This PR adds a new dropdown in the variant analysis results view, to control filtering of repositories. This allows filtering to only those repositories that have results, and is treated separate from the existing search bar which allows filtering based on the repository name.

Screenshot 2023-04-17 at 16 34 14

I've tried where at all possible to use the term "search" when referring to the existing feature of filtering based on the repository name, and used the term "filter" to refer to the new feature of filtering based on results.

The default value is "All" and therefore the default behaviour is identical to the existing behaviour. In later PRs we'll look at changing the default behaviour to "With results" and making it configurable.

I haven't added any kind of feature flag for this because I think we can implement it in shippable chunks, where each individual PR isn't too large and could be shipped on its own. If you think this PR is too large or you'd prefer to have a feature flag regardless, just let me know and we can look at that. This PR contains all the parts, including adding the UI and the backend implementation, and unit tests and a storybook story. I've tried to split up the commits as much as I could so they should all be reviewable in separately.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/test/unit-tests/variant-analysis-filter-sort.test.ts [hunk 1]
```diff
@@ -42,6 +43,60 @@ describe(matchesFilter.name, () => {
       ).toBe(matches);
     },
   );
+
+  it("returns true if filterKey is all and resultCount is positive", () => {
```
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/test/unit-tests/variant-analysis-filter-sort.test.ts:47]: I've split it up into two `describe` blocks. Let me know if that wasn't quite what you meant.

File: extensions/ql-vscode/test/unit-tests/variant-analysis-filter-sort.test.ts [hunk 2]
```diff
@@ -379,6 +434,30 @@ describe(filterAndSortRepositoriesWithResultsByName.name, () => {
...
+        filterAndSortRepositoriesWithResultsByName(repositories, {
+          ...defaultFilterSortState,
+          sortKey: SortKey.ResultsCount,
+          filterKey: FilterKey.WithResults,
+        }),
+      ).toEqual([repositories[3], repositories[2], repositories[0]]);
+    });
+  });
+
+  describe("when sort key and search and filter withResults are given", () => {
```
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/test/unit-tests/variant-analysis-filter-sort.test.ts:450]: Good point. Those definitely could be clearer 😅 

I've updated the tests that have at least three items in the list to use commas.

File: extensions/ql-vscode/src/pure/variant-analysis-filter-sort.ts [hunk 3]
```diff
@@ -105,18 +140,21 @@ export function filterAndSortRepositoriesWithResultsByName<
...
+  T extends FilterAndSortableResultWithIds,
+>(
   repositories: T[] | undefined,
   filterSortState: RepositoriesFilterSortStateWithIds | undefined,
 ): T[] | undefined {
   if (!repositories) {
     return undefined;
   }
 
+  // If repository IDs are given, then ignore the search value and filter key
```
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/src/pure/variant-analysis-filter-sort.ts:157]: Good question! I believe the answer is whenever you're operating on a list of repos selected by the user, such as when exporting results or copying a repo list to the clipboard. A user can use the tickboxes to select repos, and we pass that list into here using the repository IDs.

**Review by elenatanasoiu — APPROVED** [review]
Looks good 👍 

Thanks for splitting up the commits. It was easy to follow along.

**Pull request merged**